File: practice/633D.py
n = int(raw_input())
a = [int(x) for x in raw_input().split(' ')]
d = {}
limit = 10**9
ans = a.count(0)

for x in a:
    if not d.get(x):
        d[x] = 0
    d[x] += 1

for i in range(n):
    for j in range(n):
        if i != j and (a[i] != 0 or a[j] != 0):
            x, y = a[i], a[j]
            d[x] -= 1
            d[y] -= 1
            c = 2
            while -limit <= x and x <= limit and -limit <= y and y <= limit:
                x, y = y, x+y
                if d.get(y) and d[y] > 0:
                    d[y] -= 1
                    c += 1
                else:
                    break
            ans = max(c, ans)
            x, y = a[i], a[j]
            d[x] += 1
            d[y] += 1
            while c > 2:
                x, y = y, x+y
                d[y] += 1
                c -= 1
print ans

File: codeforces/633D.cpp
#include <bits/stdc++.h>

using namespace std;

const int limit = 1e9;

int main () {
    int n;
    cin >> n;
    vector <int> a(n);
    map <int, int> d;
    for (auto &x: a) {
        cin >> x;
        d[x] ++;
    }
    int ans = count(a.begin(), a.end(), 0);
    for (int i = 0; i < n; i ++) {
        for (int j = 0; j < n; j ++) {
            if (i != j && (a[i] != 0 || a[j] != 0)) {
                int c = 2;
                int x = a[i], y = a[j];
                d[x] --, d[y] --;
                while (abs(y) <= limit) {
                    int t = y;
                    y = x+y, x = t;
                    if (d.count(y) == 0 || d[y] == 0) {
                        break;
                    }
                    d[y] --;
                    c ++;
                }
                ans = max(ans, c);
                x = a[i], y = a[j];
                d[x] ++, d[y] ++;
                while (c > 2) {
                    int t = y;
                    y = x+y, x = t;
                    c --;
                    d[y] ++;
                }
            }
        }
    }
    cout << ans;
    return 0;
}
